src/continuous-validation/body-pair-collision.cc

```diff
@@ -142,15 +142,39 @@ value_type BodyPairCollision::collisionFreeInterval(
     bool rightIsValid = (tM > path_->timeRange().second);
     if (leftIsValid && rightIsValid) return T[0];
 
+    // Refinement step
+    // Make Nrefine times the following actions
+    //
+    //  1. Recompute the maximal velocity on interval T[2*i+0] -> Vm[2*i+1]
+    //  2. T[2*i+1] <- distanceLowerBound / Vm[2*i+1]
+    //  3. Recompute the maximal velocity on interval T[2*i+1] -> Vm[2*i+2]
+    //  4. T[2*i+2] <- distanceLowerBound / Vm[2*i+1]
+    //
+    //  The following inequalities hold:
+    //    - T [2*i+1] >=  T [2*i+0],
+    //    - Vm[2*i+1] <=  Vm[2*i],
+    //    - T [2*i+2] <=  T [2*i+1],
+    //    - Vm[2*i+2] >=  Vm[2*i+1]
+    // since
+    //    - the maximal velocity over an interval is increasing with the
+    //      size of the interval and
+    //    - distanceLowerBound / Vm is decreasing when Vm increases.
+    //
+    // Note that steps 3 and 4 are compacted into one loop in the lines
+    // below. Performing those steps an even number of times ensures that
+    // the maximal velocity used to compute the result interval T[2*i+2]
+    // is evaluated on an superset of T[2*i+2].
     for (int i = 0; i < 2 * Nrefine; ++i) {
       tm = t - (leftIsValid ? 0 : T[i]);
       tM = t + (rightIsValid ? 0 : T[i]);
       path_->velocityBound(Vb_, tm, tM);
       Vm[i + 1] = computeMaximalVelocity(Vb_);
-      T[i + 1] = distanceLowerBound / Vm[i];
+      T[i + 1] = distanceLowerBound / Vm[i + 1];
```

Comment:
Not sure how I missed this one. At some point, I remember I had a different size for `Vm` and `T` because I wasn't storing `maximalVelocity_` in `Vm`. Then I changed my mind.

What surprises me even more is that the assert passed with this bug and I observed better results. If I'm not wrong, without your changes, we have
T1 = T0
V1 >= V0
T2 = d / V1 <= d / V0 = T0
while the assert checks `T0 <= T2`. So unless `computeMaximalVelocity` always returns the same result, the assert should have failed...

Have you checked this with splines ?

Reply:
Re-reading the PR message, I see that you had an assert.

Reply:
Actually, I found this bug by compiling everything in debug mode and running the unit tests. The failure happened at https://github.com/humanoid-path-planner/hpp-core/blob/ba3f9e1ce1f0f945ce44ae11fcf9cecc36e74ada/tests/test-continuous-validation.cc#L343, when testing the splines.

Reply:
This should have been fired earlier then. This means neither me nor CI caught this.

On my side, there are two possibilities:
- I got trapped by the fact that the installed version is found before
- I thought I ran the unit tests in debug mode and I didn't.

Reply:
@nim65s, any idea how to make CI catch this in the future?
Otherwise unit tests loose a bit of their purposes.

Reply:
CI can't catch anything while it is broken. And in HPP, CI is broken most of the time, because we merge things that depend on non-released versions of dependencies.

I guess this is what happened here.

To fix this, we can either:
1. switch to a monorepo
2. run the CI in a meta repo
3. enforce the CI validation before any merge (therefore forbidding to merge things that depend on devel versions of other HPP packages)
4. get a CI that work with devel dependencies

My opinions on those propositions:
1. too much work
2. can be tried
3. would mean switching to a versioning scheme that is not always consistent between packages across HPP. I could try to do this and release one package at a time when it is needed.
4. can be tried

Reply:
Maybe this thread is not the best place to discuss this, though